esr/scheduler.go

```diff
@@ -16,132 +16,44 @@
+// Scheduler struct type with the list and three channels
 type Scheduler struct {
-	taskQueue  cleaningQueue
-	taskStream chan *cleaningTask
-	stopChan   chan struct{}
+	taskMap map[int]*time.Timer // list elements has id, timer
 }
 
-// NewScheduler creates a new scheduler
+// Returns a scheduler with an empty task map
 func NewScheduler() *Scheduler {
 	return &Scheduler{
-		taskStream: make(chan *cleaningTask),
-		stopChan:   make(chan struct{}),
+		taskMap: make(map[int]*time.Timer),
 	}
 }
 
-// AddTask adds a task to the queue with its deadline
+// AddTask adds a task to the task map and starts a timer for its job, when timer is done it runs the job in a goroutine
 func (s *Scheduler) AddTask(deadline time.Time, job func(), id int) {
-	task := &cleaningTask{
-		Deadline: deadline,
-		Job:      job,
-		Id:       id,
+	timer, exists := s.taskMap[id]
+	if exists {
+		timer.Stop()
 	}
-	s.taskStream <- task
+	t := time.AfterFunc(time.Until(deadline), job)
+	s.taskMap[id] = t
 }
 
-// RemoveTask removes a scheduled task
-func (s *Scheduler) RemoveTask(id int) bool {
-	// Search for the task with the given Id
-	for i, task := range s.taskQueue {
-		if task.Id == id {
-			// Remove the task from the queue
-			s.taskQueue = append(s.taskQueue[:i], s.taskQueue[i+1:]...)
-			heap.Init(&s.taskQueue) // Reinitialize the heap
-			return true             // Return true indicating the task was found and removed
-		}
+// RemoveTask removes a scheduled job and deletes the task from the task map
+func (s *Scheduler) RemoveTask(id int) {
+	timer, exists := s.taskMap[id]
+	if !exists {
+		return
 	}
-	return false // Return false if the task wasn't found
+	timer.Stop()
+	delete(s.taskMap, id)
 }
 
-// run is the  goroutine that cleans up expired services by continuously checking that end of validity of services
-func (s *Scheduler) run() {
-	var timer *time.Timer
-	defer s.Stop()
-	for {
-		if len(s.taskQueue) > 0 {
-			nextTask := s.taskQueue[0]
-			if timer == nil {
-				timer = time.NewTimer(time.Until(nextTask.Deadline))
-			} else {
-				timer.Reset(time.Until(nextTask.Deadline))
-			}
-		}
-
-		time.Sleep(10 * time.Millisecond) // this is used to reduce CPU consumption otherwise the go routine is a "short circuit" with no resistance
-
-		select {
-		case task := <-s.taskStream:
-			heap.Push(&s.taskQueue, task)
-			if timer == nil {
-				timer = time.NewTimer(time.Until(task.Deadline))
-			} else {
-				timer.Reset(time.Until(task.Deadline))
-			}
-		case <-func() <-chan time.Time {
-			if timer != nil {
-				return timer.C
-			}
-			return nil
-		}():
-			task := heap.Pop(&s.taskQueue).(*cleaningTask)
-			go task.Job()
-		case <-s.stopChan:
-			if timer != nil {
-				timer.Stop()
-			}
-			return
-		}
-	}
-}
-
-// Stop terminnates the scheduler
+// Stop() loops through the task map and turns off the timer for each tasks job
 func (s *Scheduler) Stop() {
-	s.stopChan <- struct{}{}
+	for _, value := range s.taskMap {
```

Comment:
"value" isn't very descriptive. In the other funcs you've called it "timer" instead, which I think fits better and remains conventional.